### PR TITLE
Refactor causal mask generation and detection for nn.transformer

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary:
* Create a private global-scope function _generate_subsequent because static class attribute member functions not supported by TorchScript resulting in torchscripting errors.
* Make TransformerEncoder and TransformerDecoder consistent w.r.t. is_causal handling by calling _detect_casual_mask
* Clarify documentation that is_causal is a hint
* Move causal mask detection into a method _detect_causal_mask 
* only accept input-size compatible causal mask as causal mask
* update _generate_subsequent_causal_mask to include factory kwargs for dtype and device:
   avoid extra copies & conversions by passing directly to torch.full.

Test Plan: sandcastle & github CICD
Continuation of #101487 (due to a tooling issue) which is a continuation-in-part of https://github.com/pytorch/pytorch/pull/98327 by @janEbert 

Differential Revision: D47427117

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -15,6 +15,21 @@
...
+        device: torch.device = torch.device('cpu'),
+        dtype: torch.dtype = torch.get_default_dtype(),
+) -> Tensor:
+    r"""Generate a square causal mask for the sequence. The masked positions are filled with float('-inf').
+        Unmasked positions are filled with float(0.0).
+    """
+    return torch.triu(
+        torch.full((sz, sz), float('-inf'), dtype=dtype, device=device),
+        diagonal=1,
+    )
```
Comment (janEbert) [inline torch/nn/modules/transformer.py:29]: Hey, in an unrelated project I just noticed that `torch.triu` cannot be applied to `bfloat16` CUDA tensors on at least PyTorch 2.0. I don't know if that changed in the meantime.

For efficiency we could do an if-query on the dtype and
- if it is `bfloat16`, move the tensor to `bfloat16` after `torch.triu` has been applied using `torch.Tensor.to`,
- otherwise (the tensor is not `bfloat16`) create it with the correct dtype immediately in `torch.full` as it is being done now.

And I guess we need to add a test that uses other dtypes to catch these things earlier.

**Comment by mikekgfb** [comment]
BF16 is a relatively new data and it is missing various support functions, esp on CPU (I recall when I tried a few years ago FP16 was also missing support on CPU for many operators.)

Is this on CPU or GPU or both?  The right course of action may be to file an issue documenting this and showing your use cases, so it might be prioritized, rather than tiptoeing around it and putting a lot of effort (and if spaghetti)! To special case some data types.

For CPU, I believe I saw messages from Intel about adding support - if you done the appropriate name in our issues you might tag that contributor. If you attach a repro, that would might help that issue.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:29]: BF16 is a relatively new data and it is missing various support functions, esp on CPU (I recall when I tried a few years ago FP16 was also missing support on CPU for many operators.)

Is this on CPU or GPU or both?  The right course of action may be to file an issue documenting this and showing your use cases, so it might be prioritized, rather than tiptoeing around it and putting a lot of effort (and if spaghetti)! To special case some data types.

For CPU, I believe I saw messages from Intel about adding support - if you done the appropriate name in our issues you might tag that contributor. If you attach a repro, that would might help that issue.

Comment (mikekgfb) [reply torch/nn/modules/transformer.py:29]: While we're on the definition of `_generate_square-ubseqent_mask`, I do wonder whether we should have a global function `F.generate_(square?)_subsequent_mask`n as a global.  The definition of creating one isn't majorly involved, but it crops up all over the place,  and having a single definition might make it easier for torch.compile() to optimize the code generation, as well as recognize comparisons of the mask against itself.

Of course, this might also done in overrides.py at the IR-level, by recognizing patterns and canonicalizing them?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -314,20 +334,8 @@ def forward(
...
-                causal_comparison = torch.triu(
-                    torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
-                ).to(mask.dtype)
-
-                if torch.equal(mask, causal_comparison):
-                    make_causal = True
-
-        is_causal = make_causal
+        seq_len = None if (src.is_nested or not hasattr(self, "seq_len_pos")) else src.size(self.seq_len_pos)
+        is_causal = _detect_is_causal_mask(mask, is_causal, seq_len)
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:338]: Do we care about preserving BC for the case where a user supplied a mask that `is_causal` but is of the wrong size?

IMO, this seems like a bug, so we don't need to provide a BC guarantee here (we can just  do a check and error if causal_mask is of the wrong size but lmk your thoughts)

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -314,20 +334,8 @@ def forward(
...
-                sz = mask.size(0)
-                causal_comparison = torch.triu(
-                    torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
-                ).to(mask.dtype)
-
-                if torch.equal(mask, causal_comparison):
-                    make_causal = True
-
-        is_causal = make_causal
+        seq_len = None if (src.is_nested or not hasattr(self, "seq_len_pos")) else src.size(self.seq_len_pos)
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:337]: This is not really a "PR review" comment but to aid my growing understanding of Transformer related nn APIs, how does the causal mask interplay with a nested `src`?

File: torch/nn/modules/transformer.py [hunk 4]
```diff
@@ -195,6 +214,7 @@ def __init__(self, encoder_layer, num_layers, norm=None, enable_nested_tensor=Tr
         # this attribute controls whether nested tensors are used
         self.use_nested_tensor = enable_nested_tensor
         self.mask_check = mask_check
+        self.seq_len_pos = 1 if encoder_layer.self_attn.batch_first else 0
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:217]: I see from the docs that that sometimes `src` can be unbatched of shape `(S, E)`. I am not sure what the value of `batch_first` will be then, just checking whether this accounts for the unbatched case

File: torch/nn/modules/transformer.py [hunk 5]
```diff
@@ -825,3 +837,45 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
...
+    make_causal = (is_causal is True)
+
+    if is_causal is None and mask is not None:
+        sz = size if size is not None else mask.size(-2)
+        causal_comparison = _generate_square_subsequent_mask(
+            sz, device=mask.device, dtype=mask.dtype)
+
+        # Do not use `torch.equal` so we handle batched masks by
+        # broadcasting the comparison.
+        if mask.size() == causal_comparison.size():
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:876]: should we also error in forward if the mask size is wrong?

File: torch/nn/modules/transformer.py [hunk 6]
```diff
@@ -150,11 +165,15 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
         return output
 
     @staticmethod
-    def generate_square_subsequent_mask(sz: int, device='cpu') -> Tensor:
-        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
+    def generate_square_subsequent_mask(
+            sz: int,
+            device: torch.device = torch.device('cpu'),
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:170]: Is there a reason why historically we always default to creating the mask on cpu here?

For context `device` and `dtype` are usually factory kwargs that default to None (e.g. [here](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L57-L58))
e.g.
```
import torch
factory_kwargs = {'device': None, 'dtype': None}
sz = 2
t = torch.full((sz, sz), float('-inf'), **factory_kwargs)
print(t.dtype, t.device)
```

 It would be nice if we could do the same here but if we make this change here though it will break BC for the edge case where the user did `torch.set_default_device('cuda')` outside the function. So I am open to leaving this like how you have it.

**Review by mikaylagawarecki — COMMENTED** [review]
Linking https://github.com/pytorch/pytorch/pull/101487 for posterity for reference to previous review comments. Looks ok but I have a few questions.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:338]: The compatibility issue is that on old models, the attribute "seq_len_pos" is not defined.  (There's gotta be a better way, but this is what we have in models all over the place -- newly introduced attributes have to be checked if they have the attribute, because data structures (I guess loaded from pickle files?) have the old definition but the new code.  

I'm happy to remove the hasattr check (actually this was the version I had before, it def looks nicer)

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:337]: `src.size()` fails on nested tensors.  you can't know what the user had before because the sequence-length is no longer a property of the batch, but each batch element hasd a length.  However, the mask would still correspond to whatever the original size was -- which you can't reconstruct from the nested tensor alone

File: torch/nn/modules/transformer.py (hunk 4 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:217]: Good catch!!!! In that case it would be 0, right?  So we have to check the dimension of the input data, and hardcode position 0 if the input is 2 dimensional.

File: torch/nn/modules/transformer.py (hunk 5 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:876]: we will error in forward -- because we don't set the is_causal flag -- so, we leave the original mask, which has the wrong shape, and at the point where the mask is used, it will fail --- like it fails today when a wrongly sized mask is passed in.

File: torch/nn/modules/transformer.py (hunk 6 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:170]: I tried permutations of thisI'm very receptive to this.  I tried with None and ran into the type checker issues -- I don't know if there's a straight-forward way.  that's why I have torch.get_default_dtype() on the dtype. Maybe we can do the same for device.

File: torch/nn/modules/transformer.py [hunk 7]
```diff
@@ -825,3 +837,45 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
...
+       may lead to reduced performance relative to what would be achievable
+       with ``is_causal=True``;
+    ``is_causal=True`` when the mask is in fact not a causal attention.mask
+       may lead to incorrect and unpredictable execution - in some scenarios,
+       a causal mask may be applied based on the hint, in other execution
+       scenarios the specified mask may be used.  The choice may not appear
+       to be deterministic, in that a number of factors like alignment,
+       hardware SKU, etc influence the decision whether to use a mask or
+       rely on the hint.
+    ``size`` if not None, check whether the mask is a causal mask of the proided size
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: fixed.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:355]: OK, rather than using a new field, I compute seq_len_pos every time, so we don't depend on an attribute that may not be present in objects created with a previous version of torch

File: torch/nn/modules/transformer.py (hunk 4 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: OK, I added _get_seq_len() function that handles all scenarios - nested tensors, unbatched tensors and batched tensors with batch first = {True, False}.

File: torch/nn/modules/transformer.py (hunk 6 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: I found a way to do this by using `torch.device(torch.C._get_default_device())` --- wayyyy to hide this functionality!

:)

File: torch/nn/modules/transformer.py [hunk 8]
```diff
@@ -385,20 +410,15 @@ def forward(self, tgt: Tensor, memory: Tensor, tgt_mask: Optional[Tensor] = None
                 the causal mask. Providing incorrect hints can result in
                 incorrect execution, including forward and backward
                 compatibility.
-            memory_is_causal: If specified, applies a causal mask as
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:388]: why did we delete this?

**Review by mikaylagawarecki — APPROVED** [review]
Looks ok, can we add the docstring for `memory_is_causal` back before merging

File: torch/nn/modules/transformer.py (hunk 8 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py]: Fixed.

**Comment by mikekgfb** [comment]
@Pytorchbot merge

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (janEbert) [reply torch/nn/modules/transformer.py:29]: Hey! I checked and the issue is only on GPU; on CPU it works. :)
I'll try to search for an issue and upvote it if I find something! If I don't, I'll follow your suggestion and open one.

I don't know enough about how this would affect TorchScript/Dynamo; shouldn't the definition in `Transformer` already be basically a singleton? I guess the major question is whether it's worth breaking backward-compatibility for the change (if you implied to remove `Transformer.generate_square_subsequent_mask`).

IMO the current way is fine, but again – I don't know how/if it affects the compiler backends. :)

Comment (janEbert) [reply torch/nn/modules/transformer.py:29]: Actually regarding bfloat16 and `torch.triu`, it seems to already be fixed upstream! https://github.com/pytorch/pytorch/pull/101414
So it's perfect that everything stayed as it is. :)

**Comment by mikekgfb** [comment]
Since you tested different versions of this function, do you want to submit them as unit test? (Ideally device be a flexible input so we can also use it to test new hardware types as support  is added)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
